### PR TITLE
[WIP Proposal] Extract bootstrapping from index.php, handle all in bootstrap.php

### PR DESCRIPTION
**Pull request by acoulton**
There is currently no clear separation between index.php and bootstrap.php - both files have to be run to bootstrap a complete environment, but loading index.php then always attempts to run either a web request or a 
minion task.

This means the bit of bootstrapping that happens in index.php has to be partially duplicated in the unittest module and also in any end-user projects that want to make the kohana environment available to other CLI or web tools (eg PHPUnit, Behat, Doctrine, etc). This causes maintenance issues if the default path structure changes, or if you want to use a non-standard directory layout. It additionally makes it more tricky than it should be to move the index.php down to an explicit web document root for security. I explained the issue more in http://dev.kohanaframework.org/issues/4777 last year.

I'd appreciate feedback/agreement on this proposal before I finish off making the required changes to unittest and minion to support it.

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -1,107 +1,14 @@
...
-}
-
-// Clean up the configuration vars
-unset($paths);
+// Bootstrap the application - modify this path if required
+require __DIR__.'/application/bootstrap.php';
 
 // If installation file exists
-if (file_exists('install'.EXT))
+if (file_exists(DOCROOT.'install'.EXT))
```
Comment (acoulton) [inline index.php:6]: The main issue I still have with this is when to include install.php. In the old version it was included by index.php immediately after directories were defined but before any other bootstrapping. It can't really happen there now, because that would prevent unit tests running amongst other things.

There's a few options:
- Duplicate the path constants in install.php (or remove reference to them) so that it can be included right at the top of index.php.
- Leave it here, so it runs after bootstrapping - but this might mean that kohana fails with errors related to some of the things the install.php is supposed to check (writable directories, for eg).
- Run it from the old location (now in bootstrap.php) but only if a new constant - say KOHANA_RUN_INSTALL_CHECKS - is defined. That constant would be set in index.php.

File: index.php [hunk 2]
```diff
@@ -1,107 +1,14 @@
...
-{
-	define('KOHANA_START_MEMORY', memory_get_usage());
+	return include DOCROOT.'install'.EXT;
 }
 
-// Bootstrap the application
-require APPPATH.'bootstrap'.EXT;
-
 // If PHP build's server API is CLI
 if (PHP_SAPI == 'cli')
```
Comment (acoulton) [inline index.php:13]: This would be removed from index.php, so that minion's task runner just loaded bootstrap and ran the task direct

File: application/bootstrap.php [hunk 3]
```diff
@@ -1,4 +1,53 @@
...
+ * deprecated notices. Disable with: E_ALL & ~E_DEPRECATED
+ */
+error_reporting(E_ALL | E_STRICT);
+
+/**
+ * The directories in which your application, modules, system and document root (web-accessible files) are located.
+ *
+ * @link http://kohanaframework.org/guide/about.install#application
+ */
+define('APPPATH', realpath(__DIR__).DIRECTORY_SEPARATOR);
```
Comment (acoulton) [inline application/bootstrap.php:23]: In the combinations I've tried, this simpler way of defining paths still supports various symlinked directory structures as opposed to the old `if (! is_dir($path) AND is_dir(DOCROOT.$path))` checks - is anyone aware of ways this might break that I've not thought of?

**Comment by lenton** [comment]
1+ I was planning on doing something similar to this myself.

In the bootstrap I think the error reporting should be moved to the top of the "Environment setup" section and all of the global definitions should be moved to the top of the "Configuration and initialization" section.

**Comment by acoulton** [comment]
@lenton agreed on the error reporting, but which global definitions are you talking about moving down to configuration and initialization?

**Comment by lenton** [comment]
@acoulton Ah, I've just realised that most of the globals are used when loading the Kohana_Core class :(

How about creating two new sections in the bootstrap and reorganising so it would look like this:

``` php
<?php

// -- Environment setup --------------------------------------------------------

error_reporting(E_ALL | E_STRICT);

...

mb_substitute_character('none');

// -- Define globals ----------------------------------------------------------

define('APPPATH', realpath(__DIR__).DIRECTORY_SEPARATOR);
define('MODPATH', realpath(__DIR__.'/../modules').DIRECTORY_SEPARATOR);
define('SYSPATH', realpath(__DIR__.'/../system').DIRECTORY_SEPARATOR);
define('DOCROOT', realpath(__DIR__.'/../').DIRECTORY_SEPARATOR);

define('EXT', '.php');

if ( ! defined('KOHANA_START_TIME'))
{
    define('KOHANA_START_TIME', microtime(TRUE));
}

if ( ! defined('KOHANA_START_MEMORY'))
{
    define('KOHANA_START_MEMORY', memory_get_usage());
}

// -- Autoloading setup --------------------------------------------------------

// Load the core Kohana class
require SYSPATH.'classes/Kohana/Core'.EXT;

if (is_file(APPPATH.'classes/Kohana'.EXT))
{
    // Application extends the core
    require APPPATH.'classes/Kohana'.EXT;
}
else
{
    // Load empty core extension
    require SYSPATH.'classes/Kohana'.EXT;
}

spl_autoload_register(array('Kohana', 'auto_load'));

//spl_autoload_register(array('Kohana', 'auto_load_lowercase'));

ini_set('unserialize_callback_func', 'spl_autoload_call');

// -- Configuration and initialization -----------------------------------------

...
```

File: index.php (hunk 1 above)
Comment (lenton) [inline index.php:6]: I'm in favour of duplicating the paths in install.php and checking for it above the bootstrap inclusion in index.php. The installation should be designed to be a one time event specifically for the default configuration of Kohana, not to be reused after customisation. This would also remove all uses of globals in the installation process which is a positive.

**Comment by acoulton** [comment]
Anyone else got an opinion on this? @enov ?

File: index.php (hunk 1 above)
Comment (mintbridge) [inline index.php:6]: Do we need it `index.php` at all? Can we not just leave it as separate file and document that to check everything is installed correctly go to `install.php`? As @lenton syas its a one time thing, so they can still just delete the file afterwards.

**Comment by enov** [comment]
:+1:  (sorry for replying late)

I like the idea of handling the bootstrapping in one file, but several ideas come to my mind:
- I think the `bootstrap.php` file should be on the `DOCROOT` besides `index.php`
- We can then have an `init.php` for the `APPLICATION` folder just like modules
- And maybe the default route should go in that `init.php`. Do you think it should be part of bootstrapping Kohana? It is something that rather belongs to your application.
- I like what @lenton said regarding duplicating the path globals for the folders in `install.php` as it is used one time and it is targeted towards newcomers.
- Regarding reorganizing, there are many ways to do that, and I suggest we leave that after that this is fully accepted and merged.
- And maybe later we can reorganize the folders further. @lenton at some point in some PR had the `cache` and `logs` files outside `application` and I liked that. Symfony 3 defines a `var` directory and we could use that to store the `cache` and `logs` inside. I think that's cool and maybe we should learn from them.Then the `application` folder resembles more and more to a module. See more here:

https://github.com/symfony/symfony-standard/issues/584#issuecomment-23148928

Also, does this PR conflict with the latest merges I made from @lenton regarding the `init_modules` new core function?

Thanks @acoulton for the great work!

**Comment by acoulton** [comment]
@enov thanks for the feedback.

> I think the bootstrap.php file should be on the DOCROOT besides index.php

Depends what you mean by DOCROOT, I think. It definitely shouldn't be directly web-accessible. It could perhaps move to / as part of moving `index.php` down to an `htdocs` or similar path by default - that is obviously a more secure setup, but it hasn't been done like that in the past to support people just dropping the whole project into a webserver (both for experimentation, and for shared hosting, as I understand it).

> We can then have an init.php for the APPLICATION folder just like modules

When would this be called? Manually from bootstrap.php, or as part of the new init_modules? Not opposed in principle, but I think it's slightly different - bootstrap.php isn't just bootstrapping kohana, it is by definition bootstrapping your application. You're bound to want to edit the standard bootstrap.php for your own projects, it might get confusing if you have changes in both a bootstrap.php and application/init.php - what's the logic for deciding what should be where?

> And maybe the default route should go in that init.php. Do you think it should be part of bootstrapping Kohana? It is something that rather belongs to your application.

I actually tend to split my routes out into a standalone routes.php - makes it much easier to use route caching.

> And maybe later we can reorganize the folders further.

Agreed - bearing in mind as with that symfony ticket we'll need to be sure end-users are aware of the need to secure any new folders (eg if logs move to DOCROOT/var) properly. It would also be good to encourage people to use /var/log where they can.

**tl:dr;** I think there's a lot of scope to tidy up bootstrap.php - it's now doing quite a lot, and mixing together definitions, configuration and actual code, which increases the friction of maintaining bootstrap.php in end-user apps. But I think that's probably for more discussion and a separate PR.

**Comment by lenton** [comment]
@enov The base routes can't be placed in the application module's init.php because they need to be set last, I got it wrong in that large PR I did. All the other ideas are great but shouldn't you post them in separate issues, it's going to get confusing if we discuss things not in the scope of the PR.

@acoulton What do you think to the bootstrap layout I suggested?

**Comment by enov** [comment]
@acoulton I think we need this to have a good unified bootstrapping. Could you kindly finalize this work and make the appropriate adjustments in `kohana/unittest` so that tests gets bootstrapped properly?

**Comment by acoulton** [comment]
@enov apologies, I've been snowed under. I don't think I'll get to this today (it also requires unittest and koharness changes and I want to do them properly per #50), and then I'm away for a week. I'll try to get on it asap when I'm back. In the meantime kohana/unittest is updated so that at least the kohana/kohana tests run for now.

**Comment by enov** [comment]
Thank you for all your hard work.

**Comment by acoulton** [comment]
@enov no prob.

**Comment by gnovaro** [comment]
I like the idea have a native file to manage routes outside the boostrap.php

**Comment by rjd22** [comment]
@acoulton @enov I think this is a good change. What still needs to be done to merge this?

**Comment by enov** [comment]
Thanks for the heads up @rjd22 -- cc @arteymix, it is here that we were discussing folder structure changes.

**Comment by acoulton** [comment]
ark. sorry guys - never did quite get back to this.

I think the outstanding work is:
- koharness needs to be updated to cope with the changes
- the unittest bootstrap needs to be redone to use `application/bootstrap`
- minion needs a proper script (probably published as a composer bin) again pointing to `application/bootstrap`
- updating the install.php so it doesn't depend on anything else

I think that's it. I _might_ get a chance to look later this week, will try to do that over getting involved in any other tickets :)

**Comment by JackEllis** [comment]
@acoulton "I actually tend to split my routes out into a standalone routes.php - makes it much easier to use route caching." - I think this should be a default.
